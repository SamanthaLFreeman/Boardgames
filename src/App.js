import React, { Component } from 'react';
import './App.css';

class App extends Component {
  componentDidMount() {
    fetch('https://www.boardgameatlas.com/api/search?order_by=popularity&ascending=false&client_id=SB1VGnDv7M')
      .then(response => response.json())
      .then(data => console.log(data.games))
      .catch(error => console.log(error))
  }
  
  render() {
    return(
      <div></div>
    )
  }

}

export default App;
