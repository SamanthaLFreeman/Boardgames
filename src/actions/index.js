export const getGames = games => ({
  type: 'GET_GAMES',
  games
})

export const getCategories = categories => ({
  type: 'GET_CATEGORIES',
  categories
})

export const addFavorite = id => ({
  type: 'ADD_FAVORITE',
  id
})

export const removeFavorite = id => ({
  type: 'REMOVE_FAVORITE',
  id
})