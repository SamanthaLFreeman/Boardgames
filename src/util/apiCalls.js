export const getPopularGames = () => {
  return fetch('https://www.boardgameatlas.com/api/search?order_by=popularity&ascending=false&client_id=igbjxf77TF')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching popular books')
      }
      return response.json()
    })
}

export const getAllCategories = () => {
  return fetch('https://www.boardgameatlas.com/api/game/categories?client_id=igbjxf77TF')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching categories')
      }
      return response.json()
    })
}