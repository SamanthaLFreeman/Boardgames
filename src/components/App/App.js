import React, { Component } from 'react';
import { getGames } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getPopularGames, getCategories } from '../../util/apiCalls';

export class App extends Component {
  componentDidMount() {
    const { getGames } = this.props;
    getPopularGames()
      .then(data => getGames(data.games))
      .catch(error => console.log(error))

    getCategories()
      .then(data => console.log(data))
      .catch(error => console.log(error))
  }
  
  render() {
    return(
      <div>
        <h1>Deck Building</h1>
      </div>
    )
  }

}

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ getGames }, dispatch)
)

export default connect(null, mapDispatchToProps)(App);
