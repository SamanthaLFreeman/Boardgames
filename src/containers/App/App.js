import React, { Component } from 'react';
import { getGames, getCategories, addFavorite, removeFavorite, addOwned, removeOwned } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getPopularGames, getAllCategories } from '../../util/apiCalls';
import CardContainer from '../CardContainer/CardContainer';
import CardDetails from '../CardDetails/CardDetails';
import Nav from '../../components/Nav/Nav';
import SearchForm from '../SearchForm/SearchForm';
import { Route, Redirect } from 'react-router-dom';


export class App extends Component {
  componentDidMount() {
    const { getGames, getCategories } = this.props;
    getPopularGames()
      .then(data => this.cleanUpGames(data.games))
      .then(data => getGames(data))
      .catch(error => console.log(error))

    getAllCategories()
      .then(data => getCategories(data.categories))
      .catch(error => console.log(error))
  }

  toggleFavorite = (favorite, bool) => {
    const { addFavorite, removeFavorite } = this.props;
    if(!bool) {
      addFavorite(favorite)
    } else {
      removeFavorite(favorite)
    }
  }

  toggleOwned = (game, bool) => {
    const { addOwned, removeOwned } = this.props;
    if(!bool) {
      addOwned(game);
    } else {
      removeOwned(game);
    }
  }

  cleanUpGames = games => {
    return games.map(game => {
      return {
        id: game.id,
        name: game.name,
        year_published: game.year_published,
        min_players: game.min_players,
        max_players: game.max_players,
        min_playtime: game.min_playtime,
        max_playtime: game.max_playtime,
        description_preview: game.description_preview,
        image_url: game.image_url,
        primary_publisher: game.primary_publisher,
        categoriesIds: game.categories,
        rules_url: game.rules_url,
        house_rules: null
      }
    })
  }
  
  render() {
    const { currentGames, ownedGames } = this.props
    return(
      <main className='app'>
        <Nav />
        <Route exact path='/searchform' render={() => <SearchForm cleanUpGames={this.cleanUpGames} /> } />
        <Route exact path='/' render={() => <CardContainer 
          toggleFavorite={this.toggleFavorite}
          toggleOwned={this.toggleOwned}
          type={'games'} />} />
        <Route exact path='/favorites' render={() => <CardContainer 
          toggleFavorite={this.toggleFavorite}
          toggleOwned={this.toggleOwned}
          type={'favorites'} />} />
        <Route exact path='/owned' render={() => <CardContainer
          toggleFavorite={this.toggleFavorite}
          toggleOwned={this.toggleOwned}
          type={'ownedGames'} />} />
        
        <Route path='/card/:id' render={({ match }) => {
          if (currentGames.length === 0) {
            return <Redirect to='/' />
          }
          let targetCard = ownedGames.find(card => card.id === match.params.id);
          if(!targetCard) {
            targetCard = currentGames.find(card => card.id === match.params.id);
          }
          return <CardDetails {...targetCard} />
        }} />
      </main>
    )
  }

}

export const mapStateToProps = state => ({
  currentGames: state.games,
  categories: state.categories,
  ownedGames: state.ownedGames
})

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ getGames, getCategories, addFavorite, removeFavorite, addOwned, removeOwned }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(App);
