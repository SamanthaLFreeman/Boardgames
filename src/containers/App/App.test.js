import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { getPopularGames, getAllCategories, searchGames, randomGame } from '../../util/apiCalls';
import { getGames, getCategories, addFavorite, removeFavorite, addOwned, removeOwned } from '../../actions';

jest.mock('../../util/apiCalls');

getPopularGames.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

searchGames.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

randomGame.mockImplementation(() => {
  return Promise.resolve({
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  })
});

getAllCategories.mockImplementation(() => {
  return Promise.resolve([{
    id: 1234,
    name: 'Zombies'
  }])
});

describe('App', () => {
  let wrapper
  const mockGames = {
    id: 42,
    name: "Catan",
    year_published: 1999,
    min_players: 2,
    max_players: 6,
    min_playtime: 30,
    max_playtime: 60,
    description_preview: "Gotta get all the sheep",
    image_url: "image",
    primary_publisher: "me",
    categories: [{ id: "1234" }],
    rules_url: "rules link"
  }
  const mockCategories = [{
    id: 1234,
    name: 'Zombies'
  }]
  const mockState = {
    games: [mockGames],
    categories: mockCategories,
    ownedGames: [mockGames],
    favorites: [mockGames]
  }
  const getGamesMock = jest.fn();
  const getCategoriesMock = jest.fn();
  const addFavoriteMock = jest.fn();
  const removeFavoriteMock = jest.fn();
  const addOwnedMock = jest.fn();
  const removeOwnedMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<App
      getGames={getGamesMock}
      getCategories={getCategoriesMock}
      addFavorite={addFavoriteMock}
      removeFavorite={removeFavoriteMock}
      addOwned={addOwnedMock}
      removeOwned={removeOwnedMock}
      />);
  });

  it('should match the snapshot with the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('mapStateToProps should grab the props it needs', () => {
    const expected = {
      currentGames: [mockGames],
      categories: mockCategories,
      ownedGames: [mockGames]
    }

    const mappedProps = mapStateToProps(mockState);
    
    expect(mappedProps).toEqual(expected);
  });

  it('it calls dispatch with the getGames action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = getGames([mockGames]);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getGames([mockGames]);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the getCategories action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = getCategories(mockCategories);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getCategories(mockCategories);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the addFavorite action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = addFavorite(mockGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addFavorite(mockGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the removeFavorite action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = removeFavorite(mockGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeFavorite(mockGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the addOwned action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = addOwned(mockGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addOwned(mockGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the removeOwned action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = removeOwned(mockGames);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeOwned(mockGames);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call addFavorite when toggleFavorite is given a false value', async () => {
    wrapper.instance().toggleFavorite(mockGames, false);
    expect(addFavoriteMock).toHaveBeenCalled();
  });

  it('should call removeFavorite when toggleFavorite is given a false value', async () => {
    wrapper.instance().toggleFavorite(mockGames, true);
    expect(removeFavoriteMock).toHaveBeenCalled();
  });

  it('should call addOwned when toggleOwned is given a false value', async () => {
    wrapper.instance().toggleOwned(mockGames, false);
    expect(addOwnedMock).toHaveBeenCalled();
  });

  it('should call removeOwned when toggleOwned is given a false value', async () => {
    wrapper.instance().toggleOwned(mockGames, true);
    expect(removeOwnedMock).toHaveBeenCalled();
  });

  it('should call getPopularGames and getAllCategories when mounting', () => {
    expect(getPopularGames).toHaveBeenCalled();
    expect(getAllCategories).toHaveBeenCalled();
  });
});
