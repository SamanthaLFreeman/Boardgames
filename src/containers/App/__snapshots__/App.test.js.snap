// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with the data passed through 1`] = `
<main>
  <Nav />
  <Connect(SearchForm)
    cleanUpGames={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    path="/card/:id"
    render={[Function]}
  />
</main>
`;
